### PR TITLE
Fix search history

### LINKED ISSUE #341: [Search] Add history support to the search page
The idea here is to:
1. Change the URL once you search something
2. Make sure the back button works

We don't need to worry about this working with the file protocol.

Ref #332

### PR DESCRIPTION
**Pull request by aphillipo**
Sorry I accidentally deleted the Pull Request in #748 it's probably a good thing because Ebert mistakes made a mess - @josevalim suggested the following people take a look:

"I am not the person to review this one. :) So let's ping @nscyclone, @dignifiedquire, @milmazz and @eksperimental."

Any ideas why these changes would cause travis to fail would be appreciated but I can't see why it's happening.

Do we support browsers that don't have the History API?

https://caniuse.com/#feat=history

Fixes #341

### CODE REVIEW

**Comment by milmazz** [comment]
@aphillipo You should include the assets created by `gulp build --type production`. I think that's why the build is failing right now.

**Comment by aphillipo** [comment]
I will fix that now @milmazz

**Comment by aphillipo** [comment]
Thanks for testing and taking the time to create video!  I knew there would be a few edge cases I didn't think of!

The 404 issue I agree with, unfortunately I can't recreate at all which is super weird.  I take it that is Chrome?  I even just tried incognito mode... How did you record your videos?

However old search results are now part of the history (by design) so surely if you search on a search page you want to keep what happened previously?  If you think we shouldn't add history on a search result page it's very simple to fix.

**Comment by nscyclone** [comment]
Thanks, @aphillipo, and yes, you're right. The 404 issue was a problem on my side. I've generated the docs from scratch and it doesn't reproduce anymore. Sorry for bothering.

Concerning the multiple search results, I agree that they should be accessible through the history, but when I press the close button I personally expect all results to be closed. At least it worked that way before.
It's quite a controversial question, it would be great if @dignifiedquire helped us with this. He has a keen eye for this kind of things.

**Comment by aphillipo** [comment]
Great!  Definitely happy to fix the remaining issue in one of the ways specified.

**Comment by josevalim** [comment]
Clicking "x" should definitely close all results. If people want to go back, then they can use the history. :)

**Comment by aphillipo** [comment]
Okay this is fixed and Ebert is now happy.  It did work first time which I find hard to believe so let me know if there are any further issues!

**Comment by josevalim** [comment]
@aphillipo super minor nitpick, can we please call the page "search.html" (all lowercase) to be consistent with everything else?

**Comment by josevalim** [comment]
@aphillipo also please rebase and regenerate the .js and .css from scratch because I did a mistake in master that went to your branch when you branched off (sorry!).

**Comment by aphillipo** [comment]
Yes to both 👍🏻  Will make changes, rebase and push again shortly.

**Comment by aphillipo** [comment]
I have completely messed up with rebase 👎

**Comment by aphillipo** [comment]
@nscyclone, @dignifiedquire, @milmazz, @eksperimental any issues with merging this?

File: assets/js/events.js [hunk 1]
```diff
@@ -103,6 +103,14 @@ function addEventListeners () {
   $('.sidebar-search .icon-search').on('click', function (e) {
     search()
   })
+
+  $(window).on('popstate', popstateHandler)
+
+  // if the search stub is refreshed or loaded perform the search.
+  if (window.location.pathname === '/search.html') {
+    var qs = getParameterByName('q')
```
Comment (milmazz) [inline assets/js/events.js]: @aphillipo could you please change this to `const` instead of `var`?

File: assets/js/search.js [hunk 2]
```diff
@@ -18,6 +18,8 @@ const $content = $('.content-inner')
 const $input = $('.sidebar-search input')
 const $sidebarItems = $('#full-list li')
 var $results
+var $oldContent
+var searchCount = 0
```
Comment (milmazz) [inline assets/js/search.js]: @aphillipo Could you please change this to `let` instead of `var`?

File: assets/js/search.js [hunk 3]
```diff
@@ -79,7 +81,23 @@ export function findIn (elements, matcher) {
   }).filter(cleaner)
 }
 
-function search (nodes, value) {
+export function getParameterByName (name, url) {
+  if (!url) url = window.location.href
+  var param = name.replace(/[\[\]]/g, '\\$&')
+  var regex = new RegExp('[?&]' + param + '(=([^&#]*)|&|#|$)')
+  var results = regex.exec(url)
```
Comment (milmazz) [inline assets/js/search.js]: @aphillipo Could you please replace this `var`s with `const`?

File: assets/js/search.js [hunk 4]
```diff
@@ -79,7 +81,23 @@ export function findIn (elements, matcher) {
...
+  if (!results) return null
+  if (!results[2]) return ''
```
Comment (milmazz) [inline assets/js/search.js]: @aphillipo Try to surround the `if` with braces even if is just one sentence. At least that's consistent with the rest of your code.

**Comment by aphillipo** [comment]
Brilliant @milmazz!  I will fix these things today.

**Comment by aphillipo** [comment]
@milmazz finally got around to fixing those things.  Thanks.

**Comment by josevalim** [comment]
@aphillipo I am re-running the build and then I will merge.

Btw, you messed something up when rebasing, we can see unrelated changes in your PR, but we will survive. :)

Thank you for your contributions!

**Comment by aphillipo** [comment]
>> Btw, you messed something up when rebasing 

I will just sort this I messed it up about 5 times so one more attempt to fix while less tired should do it!  Apologies!

**Comment by eksperimental** [comment]
Sorry @aphillipo, I am currently unavailable to test your contributions

**Comment by aphillipo** [comment]
@eksperimental no problem, I’m on holiday now anyway - when would be a good time to ping you when I’m back working?

**Comment by josevalim** [comment]
:heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:

**Pull request merged**